feature(WeekEnd Configuration)

lomamech opened the pull request:
- Create table Weekend configuration
- Implement API and Client side for week end configuration
- Implement integration test for Week end Configuration
- Implement test E2E for week end configuration

reference #2292

lomamech committed:
feature(WeekEnd Configuration)
- Create table Weekend configuration

lomamech committed:
test(Week End Configuration)
- Implement integration test for Week end Configuration

lomamech committed:
test(Week end Configuration)
- Implement test E2E for week end configuration

lomamech committed:
Resolve conflict and merge with branch Master

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -4,6 +4,7 @@
             "ACTIONS": "Actions",
             "ADD": "Add",
             "ADD_ACCOUNT_CONFIGURATION": "Add Account Configuration",
+            "ADD_WEEKEND_CONFIGURATION": "Add Week-end Configuration",
```
Just FYI: Weekend doesn't need a hyphen.  It can just be spelled "weekend".

jniles commented on client/src/modules/payroll/weekend_configuration/modals/config.modal.js:
```diff
@@ -0,0 +1,115 @@
+
+  Config.getWeekDays(vm.stateParams.id)
+    .then(function (daysConfig) {
+      daysConfig.forEach(function (object) {
```
Alternatively:
```js
Config.getWeekDays(vm.stateParams.id)
  .then(daysConfig => {
   daysConfig.forEach(object => {
     vm.weekdays.forEach(days => 
       if (days.id === object.indice) { days.checked = true; }
    });
  });

jniles commented on client/src/modules/payroll/weekend_configuration/modals/config.modal.js:
```diff
@@ -0,0 +1,115 @@
+
+    if (accountForm.$invalid || accountForm.$pristine) { return 0; }
+
+    daysChecked = vm.weekDays.filter(function (days) {
```
Alternatively:
```js
daysChecked = vm.weekdays.filter(days => days.checked)
  .map(days => days.id);
```

jniles commented on client/src/modules/payroll/weekend_configuration/templates/action.tmpl.html:
```diff
@@ -0,0 +1,27 @@
+    <span class="caret"></span>
+  </a>
+
+  <ul data-action="{{rowRenderIndex}}"  class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>
```
👍

jniles commented on test/end-to-end/weekendConfig/weekend_config.page.js:
```diff
@@ -0,0 +1,122 @@
+  /**
+   * send back the number of weekEnds in the grid
+   */
+  getWeekEndConfigCount() {
```
Do you ever use this?  Is it needed?

jniles commented on client/src/modules/payroll/weekend_configuration/configuration.service.js:
```diff
@@ -0,0 +1,32 @@
+ * Encapsulates common requests to the /weekend_config/ URL.
+ */
+function ConfigurationWeekEndService(Api, Modal, $http, util) {
+  var service = new Api('/weekend_config/');
```
Just FYI: the `PrototypeApiService` embeds the `$http` and `util` parameters.  You can get them via `API.$http` or `service.$http` for both.

jniles commented on client/src/modules/payroll/weekend_configuration/configuration.service.js:
```diff
@@ -0,0 +1,32 @@
+ * @description
+ * Encapsulates common requests to the /weekend_config/ URL.
+ */
+function ConfigurationWeekEndService(Api, Modal, $http, util) {
```
It doesn't look like Modal is ever used.  We should avoid having unused imports or functions.  The inflate our codebase and confuse people trying to read the code.

jniles commented on client/src/modules/payroll/weekend_configuration/modals/config.modal.js:
```diff
@@ -0,0 +1,115 @@
+  }
+  vm.isCreating = vm.stateParams.creating;
+
+  vm.weekDays = [
```
This looks like it should go in a service - you might want to use it again somewhere else.

jniles reviewed (commented):
Just a few comments here.

lomamech committed:
refactor(Week end Configuration)
- Use weekend instead week-end
- Update bhConstans set weekdays node
- Sanitaze configuration of week service remove not necessary function and
  use $http and util fort APIService
- Reformate the structure forEach and Filter

jniles commented on client/src/js/constants/bhConstants.js:
```diff
@@ -122,6 +122,14 @@ function constantConfig() {
       { key : 'custom_period_start', label : 'PERIODS.START', comparitor: '>', valueFilter : 'date' },
       { key : 'custom_period_end', label : 'PERIODS.END', comparitor: '<', valueFilter : 'date' },
       { key : 'limit', label : 'FORM.LABELS.LIMIT' }],
+    weekDays : [
```
This is interesting! :+1:

jniles commented on client/src/modules/payroll/weekend_configuration/configuration.service.js:
```diff
@@ -0,0 +1,32 @@
+ * @description
+ * Encapsulates common requests to the /weekend_config/ URL.
+ */
+function ConfigurationWeekEndService(Api) {
```
Nice 👍

jniles reviewed (approved):
LGTM.  The code and comments are very readable 👍 

bors r+
